### PR TITLE
Improve data stream rollover and simplify cluster metadata validation for data streams

### LINKED ISSUE #70905: Make the data stream rollover more flexible
Replace the validation that checks for conflicting indices in a data stream's backing index namespace with a more flexible data stream rollover. 

If a new write index that is created during a data stream rollover has a name that already exists in the cluster then the data stream rollover should increment the data stream's generation and perform the rollover again. This can be retried in rare case that there is another name conflict.

The current validation logic has caused problems during a cluster upgrade and currently prevents performing restoring a snapshot with data streams in a cluster that already contains these data streams.

Originated from #69669
Relates to #70861

### PR DESCRIPTION
**Pull request by danhermann**
If a naming conflict for a new write index is detected when a data stream rolls over, it will now increment its own generation until the new write index's name does not conflict with any other indices, aliases, or data streams. This allows us to remove the trappy validation around name conflicts for the backing indices of data streams.

Fixes #70905.

### CODE REVIEW

**Comment by danhermann** [comment]
@martijnvg, the main part of this change is in the first commit. The second commit moves `DataStreamTestHelper` to the same package as `DataStream` to facilitate testing so it makes a bunch of simple changes to import statements in lots of different test classes. The last three commits add or fix tests or remove obsolete tests.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java [hunk 1]
```diff
@@ -107,20 +116,26 @@ public boolean isReplicated() {
...
         }
 
         List<Index> backingIndices = new ArrayList<>(indices);
-        final String newWriteIndexName = DataStream.getDefaultBackingIndexName(getName(), getGeneration() + 1);
+        String newWriteIndexName;
+        long generation = this.generation;
+        long currentTimeMillis = timeProvider.getAsLong();
+        do {
+            newWriteIndexName = DataStream.getDefaultBackingIndexName(getName(), ++generation, currentTimeMillis);
+        } while (clusterMetadata.getIndicesLookup().containsKey(newWriteIndexName));
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:136]: 👍 simple and nicely encapsulated in this method

**Review by martijnvg — APPROVED** [review]
👍 LGTM

**Comment by danhermann** [comment]
Thanks, @martijnvg!

**Pull request merged**